// @project        The CERN Tape Archive (CTA)
// @brief          CTA Admin Command API definition
// @copyright      Copyright 2017 CERN
// @license        This program is free software: you can redistribute it and/or modify
//                 it under the terms of the GNU General Public License as published by
//                 the Free Software Foundation, either version 3 of the License, or
//                 (at your option) any later version.
//
//                 This program is distributed in the hope that it will be useful,
//                 but WITHOUT ANY WARRANTY; without even the implied warranty of
//                 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//                 GNU General Public License for more details.
//
//                 You should have received a copy of the GNU General Public License
//                 along with this program.  If not, see <http://www.gnu.org/licenses/>.

syntax  = "proto3";
package cta.admin;

//
// Command Options
//

message OptionBoolean {
  enum Key {
    BOOL_DISABLED                      =  0;
    BOOL_ENCRYPTED                     =  1;
    BOOL_FORCE                         =  2;
    BOOL_FULL                          =  3;
    BOOL_LBP                           =  4;

    // hasOption options
    BOOL_ALL                           =  5;
    BOOL_CHECK_CHECKSUM                =  6;
    BOOL_EXTENDED                      =  7;
    BOOL_SHOW_HEADER                   =  8;
    BOOL_JUSTEXPAND                    =  9;
    BOOL_JUSTREPACK                    = 10;
    BOOL_SUMMARY                       = 11;
  }

  Key key                              =  1;
  bool value                           =  2;
}

message OptionUInt64 {
  enum Key {
    UINT64_ARCHIVE_FILE_ID             =  0;
    UINT64_ARCHIVE_PRIORITY            =  1;
    UINT64_RETRIEVE_PRIORITY           =  2;
    UINT64_CAPACITY                    =  3;
    UINT64_COPY_NUMBER                 =  4;
    UINT64_FIRST_FSEQ                  =  5;
    UINT64_FILE_SIZE                   =  6;
    UINT64_LAST_FSEQ                   =  7;
    UINT64_MAX_DRIVES_ALLOWED          =  8;
    UINT64_MIN_ARCHIVE_REQUEST_AGE     =  9;
    UINT64_MIN_RETRIEVE_REQUEST_AGE    = 10;
    UINT64_NUMBER_OF_FILES             = 11;
    UINT64_PARTIAL                     = 12;
    UINT64_PARTIAL_TAPES_NUMBER        = 13;
  }

  Key key                              =  1;
  uint64 value                         =  2;
}

message OptionString {
  enum Key {
    STR_COMMENT                        =  0;
    STR_DISKID                         =  1;
    STR_DRIVE                          =  2;
    STR_ENCRYPTION_KEY                 =  3;
    STR_FILENAME                       =  4;
    STR_GROUP                          =  5;
    STR_HOSTNAME                       =  6;
    STR_INPUT                          =  7;
    STR_INSTANCE                       =  8;
    STR_LOGICAL_LIBRARY                =  9;
    STR_MOUNT_POLICY                   = 10;
    STR_OUTPUT                         = 11;
    STR_OWNER                          = 12;
    STR_PATH                           = 13;
    STR_STORAGE_CLASS                  = 14;
    STR_TAG                            = 15;
    STR_TAPE_POOL                      = 16;
    STR_USERNAME                       = 17;
    STR_VID                            = 18;
  }

  Key key                              =  1;
  string value                         =  2;
}

//
// CTA Admin Command API
//

message AdminCmd {
  enum Cmd {
    CMD_NONE                           =  0;
    CMD_ADMIN                          =  1;
    CMD_ADMINHOST                      =  2;
    CMD_ARCHIVEFILE                    =  3;
    CMD_ARCHIVEROUTE                   =  4;
    CMD_DRIVE                          =  5;
    CMD_GROUPMOUNTRULE                 =  6;
    CMD_LISTPENDINGARCHIVES            =  7;
    CMD_LISTPENDINGRETRIEVES           =  8;
    CMD_LOGICALLIBRARY                 =  9;
    CMD_MOUNTPOLICY                    = 10;
    CMD_REPACK                         = 11;
    CMD_REQUESTERMOUNTRULE             = 12;
    CMD_SHOWQUEUES                     = 13;
    CMD_SHRINK                         = 14;
    CMD_STORAGECLASS                   = 15;
    CMD_TAPE                           = 16;
    CMD_TAPEPOOL                       = 17;
    CMD_TEST                           = 18;
    CMD_VERIFY                         = 19;
  }
  enum SubCmd {
    SUBCMD_NONE                        =  0;
    SUBCMD_ADD                         =  1;
    SUBCMD_CH                          =  2;
    SUBCMD_ERR                         =  3;
    SUBCMD_LABEL                       =  4;
    SUBCMD_LS                          =  5;
    SUBCMD_RECLAIM                     =  6;
    SUBCMD_RM                          =  7;
    SUBCMD_UP                          =  8;
    SUBCMD_DOWN                        =  9;
    SUBCMD_READ                        = 10;
    SUBCMD_WRITE                       = 11;
    SUBCMD_WRITE_AUTO                  = 12;
  }

  Cmd cmd                              =  1;    //< The primary command
  SubCmd subcmd                        =  2;    //< The secondary command
  repeated OptionBoolean option_bool   =  3;    //< List of boolean options
  repeated OptionUInt64 option_uint64  =  4;    //< List of integer options
  repeated OptionString option_str     =  5;    //< List of string options
}

